### PR TITLE
feat: display an inline message in a SwiftUI app

### PR DESCRIPTION
**Pull request by ami-aman**
PR for https://linear.app/customerio/issue/MBL-402/display-an-inline-message-in-a-swiftui-app

### Problem:
As a SwiftUI developer, I want to display an inline in-app message within my SwiftUI app so that I can provide users with seamless in-app messaging experiences that are consistent with those in UIKit.

### Solution:
As UIKit views are not directly supported in SwiftUI apps hence we are using `UIViewRepresentable` to make use of `InAppMessageView` in the SwiftUI sample app. 
 
With minimal updates, add an inline message to the SwiftUI view and expect the same behaviour as a UIKit sample app.

### Demo:
- Close message and display next message in queue
- Persistent messages
- Action buttons - Close, custom
https://github.com/user-attachments/assets/2869ea46-44d5-4e7e-9b88-88fc22fcc376

- show another message
https://github.com/user-attachments/assets/153d0d79-7e84-4670-bc74-7b6be711eaf0

- deep links 
https://github.com/user-attachments/assets/0af567b6-5ab4-422e-85b4-9b377927091e

### Testing:
For manual testing, send an inline in-app message to the device
- Once it displays, check activity log events
- Tap on action button on inline message
- Check activity log for clicked callback
- Tap on close button to see activity indicator and next message in queue
- Perform same tests like on UIKit sample app

### CODE REVIEW

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift [hunk 1]
```diff
@@ -42,98 +136,110 @@ struct DashboardView: View {
...
-                            CustomEventView(close: {
-                                subscreenShown = nil
+                    Text("What would you like to test?")
+                    Group {
+                        // ---- In-app Inline View ----
+                        InAppMessageViewRepresentable(elementId: "dashboard-announcement", containerWidth: $containerWidth, heightTracker: heightTracker)
+                            // Set the height of the view based on the height tracked
+                            .frame(height: heightTracker.subviewHeight)
+                            // GeometryReader to track the size of the container
+                            .background(GeometryReader { geometry in
```
Comment (ami-aman) [inline Apps/CocoaPods-FCM/src/View/DashboardView.swift:151]: `background` adds a background color/view behind a view. In this case we are adding a clear color and when it appears it calculates the width of the view and updates it.

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift
Comment (levibostian) [inline Apps/CocoaPods-FCM/src/View/DashboardView.swift:1]: To match [the public API for the SwiftUI view](https://www.notion.so/custio/Mobile-inline-in-app-messages-technical-proposal-7efb22e2af1c40449c72f9cf73770584?pvs=4#d722b4ab9b854d84b86d744cd840d307), I was expecting that the only code added to this file was: 

```swift
InAppMessageView(elementId: "dashboard-header", onActionClick { message, actionValue, actionName in
					// See section in doc for Action button support to learn more. 
					// 
					// The function signature is the same as the existing `messageActionTaken`
					// event listener function. 
        })
```

Maybe did you forget to move the `UIViewRepresentable` code into the in-app SDK?

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift [hunk 2]
```diff
@@ -1,8 +1,102 @@
 import CioDataPipelines
+import CioMessagingInApp
 import SwiftUI
+import UIKit
```
Comment (levibostian) [inline Apps/CocoaPods-FCM/src/View/DashboardView.swift:4]:
```suggestion
```

I think this import is only here because of the `UIViewRepresentable` being in this file. 

IMO, part of the acceptance criteria of this ticket is customers only need to interact with SwiftUI framework and not need to have to import `UIKit` in their code. So before merging, I think this import should be removed.

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift [hunk 3]
```diff
@@ -1,8 +1,102 @@
...
+private var maxHeight: CGFloat = 0
+
+extension InAppMessageView {
+    var heightTracker: HeightTracker? {
+        get {
+            objc_getAssociatedObject(self, &heightTrackerKey) as? HeightTracker
+        }
+        set {
+            objc_setAssociatedObject(self, &heightTrackerKey, newValue, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
+        }
```
Comment (levibostian) [inline Apps/CocoaPods-FCM/src/View/DashboardView.swift:23]: what is the need for `objc_getAssociatedObject` and `objc_setAssociatedObject`?

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift [hunk 4]
```diff
@@ -1,8 +1,102 @@
...
+        print("Height of tracker \(String(describing: heightTracker?.subviewHeight))")
+
+        // Update maxHeight and notify heightTracker if the frame height changes
+        // The height fluctuates, increasing and then decreasing.
+        // To prevent this behavior, we constrain it to the maximum height achieved by the subview.
+        if frame.height >= maxHeight {
+            maxHeight = frame.height
+            heightTracker?.subviewHeight = frame.height
+        }
+    }
```
Comment (levibostian) [inline Apps/CocoaPods-FCM/src/View/DashboardView.swift:39]: why is this code needed?

Comment (levibostian) [reply Apps/CocoaPods-FCM/src/View/DashboardView.swift:39]: [In the demo video](https://linear.app/customerio/issue/MBL-402/display-an-inline-message-in-a-swiftui-app#comment-8a1582f9), I noticed that when the inline View renders, the height of the View instantly jumps from 0 to N. The UIKit view has an animation that animates every time the height is changed, including when the inline View is dismissed. 

Animations matching the UIKit View is part of the acceptance criteria of the ticket so we'll need to get that working.

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift [hunk 5]
```diff
@@ -1,8 +1,102 @@
...
+    @Binding var containerWidth: CGFloat
+    @ObservedObject var heightTracker: HeightTracker
+    func makeUIView(context: Context) -> InAppMessageView {
+        let inlineMessageView = InAppMessageView(elementId: elementId)
+
+        // This is optional. If set, the delegate method `onActionClick`
+        // will receive callbacks.
+        // If not set, the global method `messageActionTaken` will handle the callbacks.
+        inlineMessageView.onActionDelegate = context.coordinator
+        inlineMessageView.translatesAutoresizingMaskIntoConstraints = false
```
Comment (levibostian) [inline Apps/CocoaPods-FCM/src/View/DashboardView.swift:53]: Could you add a comment for why this is needed?

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift [hunk 6]
```diff
@@ -1,8 +1,102 @@
...
+        if frame.height >= maxHeight {
+            maxHeight = frame.height
+            heightTracker?.subviewHeight = frame.height
+        }
+    }
+}
+
+struct InAppMessageViewRepresentable: UIViewRepresentable {
+    var elementId: String
+    @Binding var containerWidth: CGFloat
```
Comment (levibostian) [inline Apps/CocoaPods-FCM/src/View/DashboardView.swift:44]: For the UIKit inline View, we assume that the width constraint is provided by the customer. I think the SwiftUI view should match that behavior where the width isn't part of constructing the SwiftUI View, but the width is instead set by the customer.

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift [hunk 7]
```diff
@@ -1,8 +1,102 @@
...
+    func updateUIView(_ uiView: InAppMessageView, context: Context) {
+        // Update the width constraint if it exists
+        if let widthConstraint = uiView.constraints.first(where: { $0.firstAttribute == .width }) {
+            widthConstraint.constant = containerWidth
+        }
+        // Update the height constraint if it exists
+        if let heightConstraint = uiView.constraints.first(where: { $0.firstAttribute == .height }) {
+            heightConstraint.constant = heightTracker.subviewHeight
+        }
+    }
```
Comment (levibostian) [inline Apps/CocoaPods-FCM/src/View/DashboardView.swift:71]: Why is this needed?

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift [hunk 8]
```diff
@@ -42,98 +136,110 @@ struct DashboardView: View {
...
+                    Group {
+                        // ---- In-app Inline View ----
+                        InAppMessageViewRepresentable(elementId: "dashboard-announcement", containerWidth: $containerWidth, heightTracker: heightTracker)
+                            // Set the height of the view based on the height tracked
+                            .frame(height: heightTracker.subviewHeight)
+                            // GeometryReader to track the size of the container
+                            .background(GeometryReader { geometry in
+                                Color.clear.onAppear {
+                                    containerWidth = geometry.size.width
+                                }
```
Comment (levibostian) [inline Apps/CocoaPods-FCM/src/View/DashboardView.swift:154]: I was expecting that this block of code [matched the public API of the SwiftUI View](https://www.notion.so/custio/Mobile-inline-in-app-messages-technical-proposal-7efb22e2af1c40449c72f9cf73770584#d722b4ab9b854d84b86d744cd840d307). Could this be fixed?

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift [hunk 9]
```diff
@@ -42,98 +136,110 @@ struct DashboardView: View {
...
                             })
-                        }
+                        // ---- In-app Inline View ----
+                        ColorButton("Send Random Event") {
```
Comment (levibostian) [inline Apps/CocoaPods-FCM/src/View/DashboardView.swift:157]: The rest of this file, was any of the code changed? Or was it just changed because of formatting the code?

**Review by levibostian — CHANGES_REQUESTED** [review]

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift
Comment (ami-aman) [reply Apps/CocoaPods-FCM/src/View/DashboardView.swift:1]: I think I missed out looking into this doc. Moved and fixed the code but I am unsure how the public API name can be same for both UIKit and SwiftUI ? I mean having same name will conflict and throw error such as invalid redeclaration.

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift (hunk 5 above)
Comment (ami-aman) [reply Apps/CocoaPods-FCM/src/View/DashboardView.swift:53]: Outdated as moved to SDK

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift (hunk 8 above)
Comment (ami-aman) [reply Apps/CocoaPods-FCM/src/View/DashboardView.swift:154]: Fixed

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift (hunk 9 above)
Comment (ami-aman) [reply Apps/CocoaPods-FCM/src/View/DashboardView.swift:157]: No other code has been changed, it could be because of the formatting that the rest of the code looks updated.

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift (hunk 7 above)
Comment (ami-aman) [reply Apps/CocoaPods-FCM/src/View/DashboardView.swift:71]: Outdated, moved to SDK.

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift (hunk 6 above)
Comment (ami-aman) [reply Apps/CocoaPods-FCM/src/View/DashboardView.swift:44]: This code is removed and outdated.

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift (hunk 2 above)
Comment (ami-aman) [reply Apps/CocoaPods-FCM/src/View/DashboardView.swift:4]: Outdated. Removed, moved to SDK.

File: Apps/CocoaPods-FCM/src/View/DashboardView.swift (hunk 3 above)
Comment (ami-aman) [reply Apps/CocoaPods-FCM/src/View/DashboardView.swift:23]: Outdated.

**Comment by levibostian** [comment]
I'll move this to draft so slack notifications step for us 🙂

**Review by levibostian — APPROVED** [review]
PR not yet ready to merge, but I am going to build on top of it with more PRs.
